File: grind75/Dec21/232.py
class MyQueue:

    # for every push, push into side stack and push back into main stack, 
LIFO + LIFO = FIFO

    def __init__(self):
        self.main_stack = []
        self.side_stack = []


    def push(self, x: int) -> None:
        for _ in range(len(self.main_stack)):
            self.side_stack.append(self.main_stack.pop())
        self.side_stack.append(x)

        for _ in range(len(self.side_stack)):
            self.main_stack.append(self.side_stack.pop())


    def pop(self) -> int:
        return self.main_stack.pop()

    def peek(self) -> int:
        return self.main_stack[-1]
        
        

    def empty(self) -> bool:
        return not self.main_stack


# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()

File: grind75/Dec21/232.go
type MyQueue struct {
    s []int
    bs []int
}


func Constructor() MyQueue {
    return MyQueue{}
}


func (this *MyQueue) Push(x int)  {
    this.s = append(this.s, x)
}


func (this *MyQueue) Pop() int {
    if len(this.bs) == 0 {
        for len(this.s) > 0 {
            this.bs = append(this.bs, this.s[len(this.s)-1])
            this.s = this.s[:len(this.s)-1]
        }
    }
    ret := this.bs[len(this.bs)-1]
    this.bs = this.bs[:len(this.bs)-1]
    return ret
}


func (this *MyQueue) Peek() int {
    if len(this.bs) == 0 {
        for len(this.s) > 0 {
            this.bs = append(this.bs, this.s[len(this.s)-1])
            this.s = this.s[:len(this.s)-1]
        }
    }
    ret := this.bs[len(this.bs)-1]
    return ret
}


func (this *MyQueue) Empty() bool {
    return len(this.bs) == 0 && len(this.s) == 0
}


/**
 * Your MyQueue object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Push(x);
 * param_2 := obj.Pop();
 * param_3 := obj.Peek();
 * param_4 := obj.Empty();
 */
